objectbox/lib/src/native/box.dart

```diff
@@ -95,6 +85,73 @@ class Box<T> {
     }
   }
 
+  /// Puts the given object in the box (persisting it) asynchronously.
+  ///
+  /// The returned future completes with an ID of a the object.  If this is a
+  /// new object (its ID property is 0), a new ID will be assigned to the object
+  /// argument, after the returned [Future] completes.
+  ///
+  /// Throws immediately if the async queue is full.
+  /// The returned future may complete with an error (throw) if the put failed
+  /// for another reason, for example a unique constraint violation. In that
+  /// case [object]'s id field remains unchanged (0 if it was a new object).
+  ///
+  /// See also [putQueued] which doesn't return a [Future] but a pre-allocated
+  /// ID immediately, even though the actual database put operation may fail.
+  Future<int> putAsync(T object, {PutMode mode = PutMode.put}) async {
+    if (_hasRelations) {
+      throw UnsupportedError('putAsync() is currently not supported on entity '
```

Comment:
it wouldn't be too hard to do but it would be much less predictable. I'd like to go with this version and possibly ease up the requirement if there are complaints. BTW objectbox-go does the same

Reply:
> I think the rule is simple enough to be predictable. 

Yeah, but there bing a rule that only triggers sometimes (based on the actual object you're inserting) means developers will only get the error occasionally, not on any put on that entity, thus may miss the cases where this happens, leading to errors in the app. If they're (un)lucky enough they will notice it later than necessary, because, you know, who reads docs...

Reply:
I'll still prep the change, see how that goes